homeassistant/components/timer.py

```diff
@@ -0,0 +1,402 @@
+        if service.service == SERVICE_START:
+            for timer in target_timers:
+                tasks.append(
+                    timer.async_start(
+                        weeks=service.data.get(
+                            ATTR_WEEKS, DEFAULT_DURATION),
+                        days=service.data.get(
+                            ATTR_DAYS, DEFAULT_DURATION),
+                        hours=service.data.get(
+                            ATTR_HOURS, DEFAULT_DURATION),
+                        minutes=service.data.get(
+                            ATTR_MINUTES, DEFAULT_DURATION),
+                        seconds=service.data.get(
+                            ATTR_SECONDS, DEFAULT_DURATION),
+                    )
+                )
+        if tasks:
+            yield from asyncio.wait(tasks, loop=hass.loop)
+
+    descriptions = yield from hass.async_add_job(
+        load_yaml_config_file, os.path.join(
+            os.path.dirname(__file__), 'services.yaml')
+    )
+
+    hass.services.async_register(
+        DOMAIN, SERVICE_START, async_handler_service,
+        descriptions[DOMAIN][SERVICE_START], SERVICE_SCHEMA_DURATION)
+    hass.services.async_register(
+        DOMAIN, SERVICE_PAUSE, async_handler_service,
+        descriptions[DOMAIN][SERVICE_PAUSE], SERVICE_SCHEMA)
+    hass.services.async_register(
+        DOMAIN, SERVICE_CANCEL, async_handler_service,
+        descriptions[DOMAIN][SERVICE_CANCEL], SERVICE_SCHEMA)
+    hass.services.async_register(
+        DOMAIN, SERVICE_FINISH, async_handler_service,
+        descriptions[DOMAIN][SERVICE_FINISH], SERVICE_SCHEMA)
+
+    yield from component.async_add_entities(entities)
+    return True
+
+
+class Timer(Entity):
+    """Representation of a timer."""
+
+    def __init__(self, hass, object_id, name, icon,
+                 weeks, days, hours, minutes, seconds):
+        """Initialize a timer."""
+        self.entity_id = ENTITY_ID_FORMAT.format(object_id)
+        self._name = name
+        self._state = STATUS_IDLE
+        self._duration = timedelta(weeks=weeks, days=days, hours=hours,
+                                   minutes=minutes, seconds=seconds)
+        self._remaining = self._duration
+        self._icon = icon
+        self._hass = hass
+        self._start = None
+        self._end = None
+        self._listener = None
+
+    @property
+    def should_poll(self):
+        """If entity should be polled."""
+        return True
+
+    @property
+    def name(self):
+        """Return name of the timer."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Return the icon to be used for this entity."""
+        return self._icon
+
+    @property
+    def state(self):
+        """Return the current value of the timer."""
+        return self._state
+
+    @property
+    def state_attributes(self):
+        """Return the state attributes."""
+        start = dt_util.as_local(self._start) if self._start else None
```

Comment:
All dates should be stored as UTC

Reply:
My intention here is to display the local time when viewed in the UI. Since I haven't done any UI yet: Can I leave it in UTC in the state and do the conversion exclusively in the UI part of this?

Reply:
Yes. Added bonus is that it can actually work 😉 The UI is not aware of the "local" time of Home Assistant.